# Maji Ndogo Water Project

In this specific audit, the objective is to **assess the integrity, and accuracy of the data stored in the database.** The data in the database was collected by a team of surveyors.  
 
In this analysis, I carried out a detailed investigation that involves analyzing the difference between the auditors report and the data collected in the field and try and Identify errors and any biasness in the data.
I will try to access the accuracy of the data, thereby ensuring that the information can be relied
 upon for decision-making and governance.


In this analysis, I investigated the differences that exist in between the auditor scores in his report and the scores collected by the surveyors.

During data collection the assigned employees re-visited several sites.

In [7]:
%reload_ext sql

In [3]:
%sql "mysql+pymysql://root:kanyi_eazy@localhost:3306/md_water_services"

In [4]:
%%sql
DROP TABLE IF EXISTS auditor_report;

CREATE TABLE auditor_report (
    location_id VARCHAR(32),
    type_of_water_source VARCHAR(64),
    true_water_source_score int DEFAULT NULL,
    statement VARCHAR(255)
);

 * mysql+pymysql://root:***@localhost:3306/md_water_services
0 rows affected.
0 rows affected.


[]

In [4]:
%%sql
SELECT *
FROM auditor_report
LIMIT 5

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('SoRu34980', 'well', 1, "Residents admired the official's commitment to enhancing urban life, praising their cooperative and inclusive approach."),
 ('AkRu08112', 'well', 3, "Villagers spoke highly of the official's dedication and genuine interest in their lives, fostering a sense of belonging and appreciation."),
 ('AkLu02044', 'river', 0, "Villagers were touched by the official's interactions, noting their humility, strong work ethic, and respectful attitude."),
 ('AkHa00421', 'well', 3, "Villagers were moved by the official's visit, praising their hard work, humility, and the profound sense of connection they fostered."),
 ('SoRu35221', 'river', 0, "A photographer's lens captures the queue, though his own struggle for water is a hidden part of the story.")]

In [6]:
%%sql
SELECT count(location_id)
FROM auditor_report
LIMIT 5

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(1620,)]

#### Intergrating the tables

From the auditors report table, there are **1620** unique sites that were revisited as seen from the cell just above. Also from the auditors table, I see a location_id, type of water source at that location, and the
 quality score of the water source, that is now independently measured. Our auditor also investigated each site a bit by speaking to a few locals.
 Their statements are also captured in his results

First, we need to compare water quality scores from the auditors table and the scores collected in the field by the surveyors. The water quality scores are stored in the water_quality table. Looking at the water_quality table, there is no column that can be used to directy join it to the auditors table. Luckily, we can use the visits table indirectly.

In [8]:
%%sql
SELECT *
FROM visits
LIMIT 5

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40),
 (3, 'AkLu01628', 'AkLu01628224', datetime.datetime(2021, 1, 1, 9, 53), 1, 0, 1),
 (4, 'AkRu03357', 'AkRu03357224', datetime.datetime(2021, 1, 1, 10, 11), 1, 28, 14)]

In [8]:
%%sql
SELECT *
FROM water_quality
LIMIT 5

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(0, 0, 1), (1, 1, 1), (2, 5, 1), (3, 10, 1), (4, 4, 1)]

In [10]:
%%sql
SELECT 
    location_id,
    true_water_source_score
FROM auditor_report
LIMIT 5

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('SoRu34980', 1),
 ('AkRu08112', 3),
 ('AkLu02044', 0),
 ('AkHa00421', 3),
 ('SoRu35221', 0)]

In [12]:
%%sql
SELECT 
    a.location_id as location,
    v.record_id ,
    a.true_water_source_score as auditor_score,
    w.subjective_quality_score as  surveyor_score
    
FROM auditor_report as a

JOIN
     visits as v
ON 
    a.location_id = v.location_id

JOIN 
    water_quality as w
ON 
    w.record_id = v.record_id
WHERE
    a.true_water_source_score = w.subjective_quality_score
AND 
    v.visit_count =1
limit 5
;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('SoRu34980', 5185, 1, 1),
 ('AkRu08112', 59367, 3, 3),
 ('AkLu02044', 37379, 0, 0),
 ('AkHa00421', 51627, 3, 3),
 ('SoRu35221', 28758, 0, 0)]

#### scores comparison

Compairing the auditors scores with the surveyors scores, we can see that the first 5 scores are similar. 
For further analysis, we need to check whether the scores are similar for all the records. To achieve this, we need to calculate whether the total 
number of record is equal to **1620** as seen above from the auditors table.

Looking at the results at the cell just below, we can see that the number scores that were common in both the auditors scores and the surviyors score were
1518. This means that out of the 1620 records in the database, 1518 are correct while the remaining were wrong.

In the folowing analysis, we want filter out the number of record that were wrong and return the names of the employees who corrected wrong information
from the field. To ascertain the credibility of the employees, we will investigate the statements made by the locals cocerning the employees.

For futher analysis, we want to calculate the total number of mistakes by each employee, and filter out those employees whose 
mistakes were below the average number of mistakes per employee.



In [13]:
%%sql
SELECT COUNT(*) as total_rows
FROM 
    auditor_report as a
JOIN 
    visits as v
ON 
    a.location_id = v.location_id
JOIN 
    water_quality as w
ON 
    w.record_id = v.record_id
WHERE 
    a.true_water_source_score = w.subjective_quality_score
AND 
    v.visit_count = 1;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(1518,)]

In [14]:
%%sql
SELECT 
    a.location_id as location,
    v.record_id ,
    ws.type_of_water_source as survey_source,
    a.type_of_water_source as auditor_source,
    a.true_water_source_score as auditor_score,
    w.subjective_quality_score as  surveyor_score
    
FROM auditor_report as a

JOIN
     visits as v
ON 
    a.location_id = v.location_id

JOIN 
    water_quality as w
ON 
    w.record_id = v.record_id
JOIN
    water_source as ws
ON
    ws.source_id = v.source_id
WHERE
    a.true_water_source_score != w.subjective_quality_score
AND 
    v.visit_count = 1 
limit 5
;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('AkRu05215', 21160, 'well', 'well', 3, 10),
 ('KiRu29290', 7938, 'shared_tap', 'shared_tap', 3, 10),
 ('KiHa22748', 43140, 'tap_in_home_broken', 'tap_in_home_broken', 9, 10),
 ('SoRu37841', 18495, 'shared_tap', 'shared_tap', 6, 10),
 ('KiRu27884', 33931, 'well', 'well', 1, 10)]

In [15]:
%%sql
SELECT 
    a.location_id as location,
    v.record_id ,
    e.employee_name,
    a.true_water_source_score as auditor_score,
    w.subjective_quality_score as  surveyor_score
    
FROM auditor_report as a

JOIN
     visits as v
ON 
    a.location_id = v.location_id

JOIN 
    water_quality as w
ON 
    w.record_id = v.record_id
JOIN 
    employee as e
ON 
    e.assigned_employee_id = v.assigned_employee_id
WHERE
    a.true_water_source_score != w.subjective_quality_score
AND 
    v.visit_count = 1 
limit 5
;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('AkRu05215', 21160, 'Rudo Imani', 3, 10),
 ('KiRu29290', 7938, 'Bello Azibo', 3, 10),
 ('KiHa22748', 43140, 'Bello Azibo', 9, 10),
 ('SoRu37841', 18495, 'Rudo Imani', 6, 10),
 ('KiRu27884', 33931, 'Bello Azibo', 1, 10)]

In [16]:
%%sql
with Incorrect_records as(                      
SELECT 
    a.location_id as location,
    v.record_id ,
    e.employee_name,
    a.true_water_source_score as auditor_score,
    w.subjective_quality_score as  surveyor_score
    
FROM auditor_report as a

JOIN
     visits as v
ON 
    a.location_id = v.location_id

JOIN 
    water_quality as w
ON 
    w.record_id = v.record_id
JOIN 
    employee as e
ON 
    e.assigned_employee_id = v.assigned_employee_id
WHERE
    a.true_water_source_score != w.subjective_quality_score
AND 
    v.visit_count = 1 
),

error_record as (
SELECT 
    i.employee_name,
    count(*) as number_of_mistakes
FROM Incorrect_records as i
GROUP BY employee_name
ORDER BY number_of_mistakes DESC
)


SELECT
 r.employee_name,
 r.number_of_mistakes
 FROM
 error_record as r
 WHERE
 r.number_of_mistakes > (
     SELECT 
        avg(e.number_of_mistakes) as avg_error_count_per_empl
    FROM error_record as e)
;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
4 rows affected.


[('Bello Azibo', 26),
 ('Malachi Mavuso', 21),
 ('Zuriel Matembo', 17),
 ('Lalitha Kaburi', 7)]

In [ ]:
%%sql
CREATE VIEW Incorrect_records as (
SELECT
     a.location_id,
     v.record_id,
     e.employee_name,
     a.true_water_source_score AS auditor_score,
     wq.subjective_quality_score AS surveyor_score,
     a.statement AS statement
 FROM
     auditor_report as a
 JOIN
     visits as v
 ON 
     a.location_id = v.location_id
 JOIN
     water_quality AS wq
 ON 
    v.record_id = wq.record_id
 JOIN
     employee as e
 ON 
     e.assigned_employee_id = v.assigned_employee_id
 WHERE
     v.visit_count =1
 AND 
    a.true_water_source_score != wq.subjective_quality_score)
;

In [17]:
%%sql
SELECT * FROM Incorrect_records LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('AkRu05215', 21160, 'Rudo Imani', 3, 10, "Villagers admired the official's visit for its respectful interactions, hard work, and genuine concern."),
 ('KiRu29290', 7938, 'Bello Azibo', 3, 10, 'A young artist sketches the faces in the queue, capturing the weariness of daily hours spent waiting for water.'),
 ('KiHa22748', 43140, 'Bello Azibo', 9, 10, "A young girl's hopeful eyes are clouded by mistrust, her innocence tarnished by the corrupt system."),
 ('SoRu37841', 18495, 'Rudo Imani', 6, 10, "The official's respectful and diligent presence was met with heartfelt appreciation, creating a sense of closeness with the villagers."),
 ('KiRu27884', 33931, 'Bello Azibo', 1, 10, "A traditional healer's empathy turns to bitterness, knowing that corrupt practices harm her community.")]

In [27]:
%%sql
with error_record as (
SELECT 
    employee_name,
    count(employee_name) as number_of_mistakes
FROM 
    Incorrect_records 
GROUP BY employee_name
ORDER BY number_of_mistakes DESC
),

suspects_list as
(SELECT
 r.employee_name,
 r.number_of_mistakes
 FROM
 error_record as r
 WHERE
 r.number_of_mistakes > (
     SELECT 
        avg(e.number_of_mistakes) as avg_error_count_per_empl
    FROM error_record as e)
)

SELECT * FROM suspects_list
;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
4 rows affected.


[('Bello Azibo', 26),
 ('Malachi Mavuso', 21),
 ('Zuriel Matembo', 17),
 ('Lalitha Kaburi', 7)]

#### conclusion

Looking at the table just above, the number of employees who made mistake greater than the average number of mistakes were 5. 
These employees are listed below together with the number of mistakes each empoyee did. 

1. **Bello Azibo**	 **26**

2. **Malachi Mavuso	21**

3. **Zuriel Matembo	17**

4. **Lalitha Kaburi	7**

If we were to further invetigatethe comments made by the locals concerning the employeee, it is clear that this employees were corrupt and immediae action should be taken to deal with them.
    
    

In [9]:
%%sql
SELECT 
    location_id,
    employee_name,
    statement
FROM 
    Incorrect_records
WHERE 
    employee_name 
in ("Bello Azibo","Malachi Mavuso","Zuriel Matembo","Lalitha Kaburi")
AND statement 
LIKE "%cash%"
limit 5
;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('SoRu38331', 'Zuriel Matembo', 'An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.'),
 ('AmAm09607', 'Malachi Mavuso', 'Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.'),
 ('KiIs23853', 'Bello Azibo', "Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The mention of cash changing hands further tainted their perception."),
 ('HaSe21323', 'Bello Azibo', 'Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.'),
 ('AkRu05880', 'Zuriel Matembo', "Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The allusion to cash changing hands deepened their skepticism.")]